### PR TITLE
Add support for wait strategies

### LINKED ISSUE #123: Add support for Wait Strategies for generic containers
### Feature description

It would be very useful to have support for customizing the criteria for when a generic container is ready for use, like documented here: https://www.testcontainers.org/features/startup_and_waits/

An example would be something like this in `application.yml`

```
test-resources:
  containers:
    custom-container:
      image-name: my-image
      hostnames:
        - my.host
      exposed-ports:
        - my.port: 9876
      wait-strategy:
        - log-message: ".*Ready to accept connections.*\\n"
```

### LINKED ISSUE #136: Allow additional waiting strategy for the Database to be available
### Feature description

Unfortunately, there is something different with timing when Rancher Desktop is used which makes Postgres (probably Oracle) container tests fail because it's reported as ready(it uses `LogMessageWaitStrategy`), but it still takes some time for the port to be exposed or something...

It would be nice to have some additional availability checks that can be enabled. I can see two options:

1. Await for successful SQL query, that is the default Testcontainers check:
https://github.com/testcontainers/testcontainers-java/blob/7ba6b1c518d8f181b5831229977c260dbff803dc/modules/jdbc/src/main/java/org/testcontainers/containers/JdbcDatabaseContainer.java#L143-L175
Unfortunately, it's overridden by Postgres and Oracle with `LogMessageWaitStrategy`.
This check might be the most bulletproof but might need a custom configuration of the SQL check query.

2. Wait for the port to be available. We can probably reuse Testcontainer's `HostPortWaitStrategy`.

### PR DESCRIPTION
**Pull request by melix**
Fixes #123
Fixes #136

### CODE REVIEW

File: test-resources-testcontainers/src/main/java/io/micronaut/testresources/testcontainers/TestContainerMetadataSupport.java [hunk 1]
```diff
@@ -225,6 +238,82 @@ public String getValue() {
             .orElse(Collections.emptyMap());
     }
 
+    private static WaitStrategy extractWaitStrategyFrom(String prefix, Map<String, Object> testResourcesConfiguration) {
```
Comment (alvarosanchez) [inline test-resources-testcontainers/src/main/java/io/micronaut/testresources/testcontainers/TestContainerMetadataSupport.java:241]: Why don't you use a configuration class and let Micronaut parse and validate it for you?

Comment (dstepanov) [reply test-resources-testcontainers/src/main/java/io/micronaut/testresources/testcontainers/TestContainerMetadataSupport.java:241]: There is no bean context

**Comment by dstepanov** [comment]
Is it supposed to work with any containers? JDBC once included?

File: src/main/docs/guide/modules-testcontainers.adoc [hunk 2]
```diff
@@ -147,3 +147,33 @@ test-resources:
...
+
+- `log` strategy with the following properties:
+    - `regex`: the regular expression to match
+    - `times`: the number of times the regex should be matched
+- `http` strategy with the following properties:
+    - `path`: the path to check
+    - `tls`: if TLS should be enabled
+    - `port`: the port to listen on
+    - `status-code`: the list of expected status codes
+- `port` strategy (this is the default strategy)
```
Comment (dstepanov) [inline src/main/docs/guide/modules-testcontainers.adoc:178]: Do you mean it’s a default strategy for the custom containers? Most of the JDBC containers have SQL teat query except for Oracle and Postgres which come with the log check.

Comment (melix) [reply src/main/docs/guide/modules-testcontainers.adoc:178]: It's the default strategy that is applied by test containers on generic containers.

File: test-resources-testcontainers/src/main/java/io/micronaut/testresources/testcontainers/TestContainerMetadataSupport.java (hunk 1 above)
Comment (melix) [reply test-resources-testcontainers/src/main/java/io/micronaut/testresources/testcontainers/TestContainerMetadataSupport.java:241]: Correct: test resources are setup _before_ the bean context is available.

**Comment by melix** [comment]
> Is it supposed to work with any containers? JDBC once included?

Yes, it works with any container, in the sense that it sets the wait strategy. However, it doesn't do what you'd expect, which is to override, for example, the `SELECT 1` query. I am trying to figure out why it is.

**Comment by melix** [comment]
@dstepanov https://github.com/testcontainers/testcontainers-java/blob/96acaf02721c5dbf543ff2b166366a190df57cfa/core/src/main/java/org/testcontainers/containers/GenericContainer.java#L950-L955

This is where the wait strategy is applied. The JDBC containers _override_ this method and basically ignore any custom wait strategy, so this won't let you do what you want.

**Comment by melix** [comment]
Actually in your case it would, since you'd replace the log strategy with a port strategy, which is supported by the Postgres container. It's just that _in general_ we cannot always say that the configured strategy will be used.

**Pull request merged**